File: C++/src-lab2/lab2p2.cpp
#include <iostream>
#include <vector>

#define CEIL(a, b) ((a) / (b) + ((a) % (b) ? 1 : 0))

/* Modelam o fractie ca fiind o pereche (numarator, numitor). */
typedef std::pair<long long, long long> Fractie;

std::vector<Fractie> descompunere_fractii_egiptene(long long x, long long y)
{
  std::vector<Fractie> v;

  /* TODO: Puneti in v un sir de fractii care sa reprezinte descompunerea in
   * fractii egiptene a lui x/y. */

  return v;
}

int main()
{
  /* Citim x si y. */
  long long x, y;
  std::cin >> x >> y;

  /* Calculam descompunerea in fractii egiptene. */
  std::vector<Fractie> v;
  v = descompunere_fractii_egiptene(x, y);

  /* Si o afisam. */
  std::cout << x << "/" << y << " = ";
  for (unsigned int i = 0; i < v.size(); ++i) {
    std::cout << (i > 0 ? " + " : "") << v[i].first << "/" << v[i].second;
  }
  std::cout << std::endl;

  return 0;
}

File: Java/src-lab2/Lab2p2.java
import java.util.Scanner;
import java.util.Vector;

public class Lab2p2 
{
  public static void main(String[] args)
  {
    /* Citim x si y. */
    Scanner inputScanner = new Scanner(System.in);
    long x = inputScanner.nextLong();
    long y = inputScanner.nextLong();

    /* Calculam descompunerea in fractii egiptene. */
    Vector<Fractie> v = descompunere_fractii_egiptene(x, y);

    /* Si o afisam. */
    System.out.println(v);
  }

  private static long ceil(long a, long b)
  {
    return a / b + (a % b > 0 ? 1 : 0); 
  }

  private static Vector<Fractie> descompunere_fractii_egiptene(long x, long y) 
  {
    Vector<Fractie> v = new Vector<Fractie>();
    /* Cat timp nu am ajuns cu descompunerea la 0, continuam sa alegem cea mai
     * mare fractie de forma 1/n care este strict mai mica decat x/y, o adaugam la
     * suma si incecam sa descompunem mai departe diferenta (x/y - 1/n).
     *
     * Se observa ca acea fractie este mereu 1/(CEIL(y/x)).
     */
    do 
    {
      v.add(new Fractie(1, ceil(y, x)));
      long newx = (x - y % x) % x;
      long newy = y * ceil(y, x);
      x = newx;
      y = newy;
    } while (x != 0);

    return v;
  }
}

class Fractie
{
  public long numarator;
  public long numitor;

  public Fractie(long numarator, long numitor)
  {
    this.numarator = numarator;
    this.numitor = numitor;
  }

  public String toString()
  {
    return numarator + " / " + numitor;
  }
}
